File: YelpSwiftApp/Models/YelpClient.m
//
//  YelpClient.m
//  Yelp
//
//  Created by Timothy Lee on 3/21/14.
//  Copyright (c) 2014 codepath. All rights reserved.
//

#import "YelpClient.h"
#import "YelpBusiness.h"

// You can register for Yelp API keys here: http://www.yelp.com/developers/manage_api_keys
NSString * const kYelpConsumerKey = @"cJL5QRpvSjF3k0UNbdhvSg";
NSString * const kYelpConsumerSecret = @"lEaBTbu1MGc9tRUFwE9vqkX6TvU";
NSString * const kYelpToken = @"Tj8hWSjzNc3qw-l-9HP3N3qwYpXbf3Iy";
NSString * const kYelpTokenSecret = @"QMHf7YrrXKmdOTXeowtytqRjS1c";

@interface YelpClient ()

@end

@implementation YelpClient

// Singleton
+ (instancetype)sharedInstance {
    static YelpClient *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[YelpClient alloc] init];
    });
    return instance;
}

- (instancetype)init {
    NSURL *baseURL = [NSURL URLWithString:@"https://api.yelp.com/v2/"];
    if (self = [super initWithBaseURL:baseURL consumerKey:kYelpConsumerKey consumerSecret:kYelpConsumerSecret]) {
        
        BDBOAuth1Credential *token = [BDBOAuth1Credential credentialWithToken:kYelpToken
                                                                       secret:kYelpTokenSecret
                                                                   expiration:nil];
        [self.requestSerializer saveAccessToken:token];
    }
    return self;
}

- (AFHTTPRequestOperation *)searchWithTerm:(NSString *)term
                                completion:(void (^)(NSArray *businesses, NSError *error))completion {
    
    return [self searchWithTerm:term
                       sortMode:YelpSortModeBestMatched
                     categories:nil
                          deals:NO
                     completion:completion];
}

- (AFHTTPRequestOperation *)searchWithTerm:(NSString *)term
                                  sortMode:(YelpSortMode)sortMode
                                categories:(NSArray *)categories
                                     deals:(BOOL)hasDeal
                                completion:(void (^)(NSArray *businesses, NSError *error))completion {
    
    // For additional parameters, see http://www.yelp.com/developers/documentation/v2/search_api
    NSMutableDictionary *parameters = [@{@"term": term,
                                         @"ll" : @"37.774866,-122.394556",
                                         @"sort": [NSNumber numberWithInt:sortMode]}
                                       mutableCopy];
    
    if (categories && categories.count > 0) {
        parameters[@"category_filter"] = [categories componentsJoinedByString:@","];
    }
    
    if (hasDeal) {
        parameters[@"deals_filter"] = [NSNumber numberWithBool:hasDeal];
    }
    
    NSLog(@"%@", parameters);
    
    return [self GET:@"search"
          parameters:parameters
             success:^(AFHTTPRequestOperation * _Nonnull operation, id  _Nonnull responseObject) {
                 
                 NSArray *businesses = responseObject[@"businesses"];
                 completion([YelpBusiness businessesFromJsonArray:businesses], nil);
                 
             } failure:^(AFHTTPRequestOperation * _Nonnull operation, NSError * _Nonnull error) {
                 completion(nil, error);
             }];
}



- (AFHTTPRequestOperation *)searchWithParams:(NSDictionary *)params
                                  completion:(void (^)(NSArray *businesses, NSError *error))completion {
    
    NSLog(@"%@", params);
//    NSMutableDictionary *parameters = [@{@"term": "restaurants",
//                                         @"ll" : @"37.774866,-122.394556",
//                                         @"sort": [NSNumber numberWithInt:sortMode]}
//                                       mutableCopy];

    return [self GET:@"search"
          parameters:params
             success:^(AFHTTPRequestOperation * _Nonnull operation, id  _Nonnull responseObject) {
                 
                 NSArray *businesses = responseObject[@"businesses"];
                 completion([YelpBusiness businessesFromJsonArray:businesses], nil);
                 
             } failure:^(AFHTTPRequestOperation * _Nonnull operation, NSError * _Nonnull error) {
                 completion(nil, error);
             }];
}

@end

File: YelpSwiftApp/Models/YelpClient.swift
//
//  YelpClient.swift
//  YelpSwiftApp
//
//  Created by Scott Richards on 5/20/16.
//  Copyright © 2016 Baby Center. All rights reserved.
//

import Foundation


class YelpClient : BDBOAuth1RequestOperationManager
{
    enum SortOrder {
        case BestMatched
        case ModeDistance
        case HighestRated
    }
    
    private static var _sharedInstance : YelpClient? = nil
    static var sharedInstance : YelpClient {
        get {
            if let _sharedInstance = _sharedInstance {
                return _sharedInstance
            } else {
                _sharedInstance = YelpClient()
                return _sharedInstance!
            }
        }
    }
    
    struct keys {
        static let kYelpConsumerKey = "cJL5QRpvSjF3k0UNbdhvSg"
        static let kYelpConsumerSecret = "lEaBTbu1MGc9tRUFwE9vqkX6TvU"
        static let kYelpToken = "Tj8hWSjzNc3qw-l-9HP3N3qwYpXbf3Iy"
        static let kYelpTokenSecret = "QMHf7YrrXKmdOTXeowtytqRjS1c"
    }
    
    // initialize with BaseURL pass Token and Keys
    init() {
        let baseURL : NSURL? = NSURL(string: "https://api.yelp.com/v2/")
        super.init(baseURL: baseURL, consumerKey: keys.kYelpConsumerKey, consumerSecret: keys.kYelpConsumerSecret)
        let token : BDBOAuth1Credential = BDBOAuth1Credential(token: keys.kYelpToken, secret: keys.kYelpTokenSecret, expiration: nil)
        self.requestSerializer.saveAccessToken(token)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func searchWithParams(params : NSDictionary, offset: UInt = 0, limit: UInt = 20, completionHandler: (businesses : [YelpBusiness]?,error : NSError?) -> Void) {
        let parameters : NSMutableDictionary = NSMutableDictionary(dictionary: params)
        if let offsetStr : String = String(format: "%d", offset) {
            parameters.setObject(offsetStr, forKey:"offset")
        }
        if let limitStr : String = "\(limit)" {
            parameters.setObject(limitStr, forKey:"limit")
        }
        for param in parameters {
            print("param: \(param)")
        }
        self.GET("search", parameters: parameters,
                 success: {
                    (operation : AFHTTPRequestOperation, result : AnyObject) in
                    print("success")
                    let businesses : [AnyObject] = result["businesses"] as! [AnyObject]
                    completionHandler(businesses: YelpBusiness.businessesFromJsonArray(businesses) as [YelpBusiness], error: nil)
            },
                 
                 failure: {
                    (operation : AFHTTPRequestOperation?, error : NSError) in
                    print("failure")
                    completionHandler(businesses: nil, error: error)
            }
        )
    }
    
    func searchWithTerm(term : String?, offset: UInt = 0, limit: UInt = 20, completion : (businesses:[YelpBusiness]?, error : NSError?) -> Void) {
        let params : NSMutableDictionary? = ["ll":"37.774866,-122.394556"]
        if let term = term {
            params?.setObject(term, forKey:"term")
        }
        if let offsetStr : String = String(format: "%d", offset) {
            params?.setObject(offsetStr, forKey:"offset")
        }
        if let limitStr : String = "\(limit)" {
            params?.setObject(limitStr, forKey:"limit")
        }
        let stringNumber = "1234"
        if let  numberFromString : Int = Int(stringNumber) {
          print(numberFromString)
        }
        let appDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        if (appDelegate.userLocation.foundLocation == true) {
            params?.setObject("\(appDelegate.userLocation.latitude),\(appDelegate.userLocation.longitude)", forKey: "ll")
        }
        self.searchWithParams(params!, completionHandler: completion)
    }
    
//    func searchWithTerm(term : String?, sortMode : SortOrder? = .BestMatched, completion : ([AnyObject]?, error : NSError?)) {
//        
//    }
    
}